src/sentry/api/endpoints/user_authenticator_enroll.py

```diff
@@ -0,0 +1,195 @@
+        max_length=20,
+    )
+    otp = serializers.CharField(
+        label='One-time password',
+        help_text='Code from authenticator',
+        required=False,
+        max_length=20
+    )
+
+
+class U2fRestSerializer(serializers.Serializer):
+    deviceName = serializers.CharField(
+        label='Device name',
+        required=False,
+        max_length=60,
+        default=lambda: petname.Generate(2, ' ', letters=10).title(),
+    )
+
+serializer_map = {
+    'totp': TotpRestSerializer,
+    'sms': SmsRestSerializer,
+    'u2f': U2fRestSerializer,
+}
+
+
+def get_serializer_field_metadata(serializer, fields=None):
+    """Returns field metadata for serializer"""
+    meta = []
+    for field in serializer.base_fields:
+        if fields is None or field in fields:
+            serialized_field = dict(serializer.base_fields[field].metadata())
+            serialized_field['name'] = field
+            serialized_field['defaultValue'] = serializer.base_fields[field].get_default_value()
+            meta.append(serialized_field)
+
+    return meta
+
+
+class UserAuthenticatorEnrollEndpoint(UserEndpoint):
+    permission_classes = (IsAuthenticated, UserPermission, )
+
+    @sudo_required
+    def get(self, request, user, interface_id):
+        """
+        Get Authenticator Interface
+        ```````````````````````````
+
+        Retrieves authenticator interface details for user depending on user enrollment status
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        interface = Authenticator.objects.get_interface(user, interface_id)
+
+        # Not all interfaces allow multi enrollment
+        if interface.is_enrolled and not interface.allow_multi_enrollment:
+            return Response(status=status.HTTP_400_BAD_REQUEST)
+
+        # User is not enrolled in auth interface:
+        # - display configuration form
+        response = serialize(interface)
+        response['form'] = get_serializer_field_metadata(
+            serializer_map[interface_id]
+        )
+
+        # U2fInterface has no 'secret' attribute
+        try:
+            response['secret'] = interface.secret
+        except AttributeError:
+            pass
+
+        if interface_id == 'totp':
+            response['qrcode'] = interface.get_provision_qrcode(user.email)
+
+        if interface_id == 'u2f':
+            response['challenge'] = interface.start_enrollment()
+
+        return Response(response)
+
+    @sudo_required
+    def post(self, request, user, interface_id):
+        """
+        Enroll in authenticator interface
+        `````````````````````````````````
+
+        :pparam string user_id: user id or "me" for current user
+        :pparam string interface_id: interface id
+
+        :auth: required
+        """
+
+        # start activation
+        serializer_cls = serializer_map.get(interface_id, None)
```

Comment:
you shouldn't need `None` here (`.get` should return that as the default if no default is provided)